docs/source/basic/library.rst

```diff
@@ -115,6 +115,68 @@ The ``alpaka::getWorkDiv`` and the ``alpaka::getIdx`` functions both return a ve
 They are parametrized by the origin of the calculation as well as the unit in which the values are calculated.
 For example, ``alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc)`` returns a vector with the extents of the grid in units of threads.
 
+Memory fences
+`````````````
+
+**Note**: Memory fences should not be mistaken for synchronization functions between threads. They solely enforce the
+ordering of certain memory operations (see below) and how other threads observe this order. If you need to rely on the
+results of memory operations being visible to other threads please use ``alpaka::syncBlockThreads`` or atomic functions
+instead.
+
+The ``alpaka::mem_fence`` function can be used inside an alpaka kernel to issue a memory fence instruction. This
+guarantees the following **for the local thread**  and regardless of global or shared memory:
+
+  * All loads that occur before the fence will happen before all loads occuring after the fence, i.e. no *LoadLoad*
+    reordering.
+  * All stores that occur before the fence will happen before all stores occuring after the fence, i.e. no *StoreStore*
+    reordering.
+  * The order of stores will be visible to other threads inside the scope (but not necessarily their results).
+
+**Note**: ``alpaka::mem_fence`` does not guarantee that there will be no *LoadStore* reordering. Depending on the
+back-end, loads occurring before the fence may still be reordered with stores occurring after the fence.
+
+Memory fences can be issued on the block level (``alpaka::memory_scope::Block``) and the device level
+(``alpaka::memory_scope::Device``). Depending on the memory scope, the *StoreStore* order will be visible to other
+threads in the same block or the whole device.
+
+Some accelerators (like GPUs) follow weaker cache coherency rules than x86 CPUs. In order to avoid storing to (or loading
+from) a cache or register it is necessary to prefix all observed buffers with `ALPAKA_DEVICE_VOLATILE`. This enforces
+that all loads / stores access the actual global / shared memory location.
+
+Example:
+
+.. code-block:: cpp
+
+    /* Initial values: 
+     * vars[0] = 1
+     * vars[1] = 2
+     */
+    template<typename TAcc>
+    ALPAKA_FN_ACC auto operator()(TAcc const& acc, bool* success, ALPAKA_DEVICE_VOLATILE int* vars) const -> void
+    {
+        auto const idx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u];
+
+        // Global thread 0 is producer
+        if(idx == 0)
+        {
+            vars[0] = 10;
+            alpaka::mem_fence(acc, alpaka::memory_scope::Device{});
+            vars[1] = 20;
+        }
+
+        auto const b = vars[1];
+        alpaka::mem_fence(acc, alpaka::memory_scope::Device{});
+        auto const a = vars[0];
+
+        /* Possible results at this point:
+         * a == 1 && b == 2
+         * a == 10 && b == 2
+         * a == 10 && b == 20
+         *
+         * but NOT:
+         * a == 1 && b == 20
+         */
+    }
```

Comment:
I am starting to think that this code is undefined in C++, as it is missing an atomic operation. See: https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence

> Fence-fence synchronization
> A release fence FA in thread A synchronizes-with an acquire fence FB in thread B, if
> There exists an atomic object M,

This code does not have an atomic, and volatile does replace this behavior.

Reply:
Ugh, I thought I could do this without having to call another atomic operation (on godbolt the correct CPU instructions are generated by `std::atomic_thread_fence`), but you're right.